### PR TITLE
MAINT: cleanups to _iotools.StringConverter

### PR DESCRIPTION
**Pull request by eric-wieser**
Some preliminary cleanups while exploring trying to insert a `DeprecationWarning` for surprising behavior.

### CODE REVIEW

File: numpy/lib/_iotools.py [hunk 1]
```diff
@@ -621,36 +622,26 @@ def __init__(self, dtype_or_func=None, default=None, missing_values=None,
...
-                _status = 0
+
+            # find the best match in our mapper
+            try:
+                self._status, (_, func, default_def) = self._find_map_entry(dtype)
+            except LookupError:
+                # no match
                 self.default = default
-            self._status = _status
+                _, func, _ = self._mapper[-1]
```
Comment (eric-wieser) [inline numpy/lib/_iotools.py:632]: @seberg, simplifying this logic is how I found #16189

**Comment by seberg** [comment]
@eric-wieser lets do this (will review some other time probably), but thought you should know that Warren is looking into hopefully replacing the whole text parsing. That is a bit of a long game obviously, but in case it matters for priorities.

**Comment by eric-wieser** [comment]
Just to give a bit more background here, the motivation was to eventually try and deprecate the surprising behavior that #15791 raises, that `genfromtxt` will dangerously replace missing values with `-1` even if not asked too. IMO that should at least emit a warning unless the user explicitly asks for missing values to be replaced.

**Comment by seberg** [comment]
Ah, deprecations are always good... IMO, also helps with future changes, since replicating weird behaviour is usually unnecessarily hard...

File: numpy/lib/_iotools.py [hunk 2]
```diff
@@ -586,9 +574,22 @@ def upgrade_mapper(cls, func, default=None):
...
+    def _find_map_entry(cls, dtype):
+        # if a converter for the specific dtype is available use that
+        for i, (deftype, func, default_def) in enumerate(cls._mapper):
+            if dtype.type == deftype:
+                return i, (deftype, func, default_def)
+
+        # otherwise find an inexact match
+        for i, (deftype, func, default_def) in enumerate(cls._mapper):
+            if np.issubdtype(dtype.type, deftype):
+                return i, (deftype, func, default_def)
```
Comment (anirudh2290) [inline numpy/lib/_iotools.py:590]: doesn't look like we need deftype for the caller, do we need to return this ?

Comment (eric-wieser) [reply numpy/lib/_iotools.py:590]: It's not used, but it means the return type matches the entries in `_mapper`, which I thought made it clearer.

**Comment by seberg** [comment]
Thanks Eric and Anirudh for the review. Looks correct to me as well, so going to put it in. Anirudh comment may make sense, but just as well for a later cleanup.

**Pull request merged**